Plugins/GeoModel/src/SplitModulesByRadius.cpp

```
@@ -0,0 +1,84 @@
+
+  if (annulusBounds == nullptr) {
+    ACTS_DEBUG("Not annulus bounds");
+    return std::nullopt;
+  }
+
+  std::optional<std::vector<std::pair<
+      std::shared_ptr<Surface>, std::shared_ptr<GeoModelDetectorElement>>>>
+      result;
+
+  for (auto [design, radii] : m_designs) {
+    if ((std::abs(radii.front() - annulusBounds->rMin()) > m_tolerance) ||
+        (std::abs(radii.back() - annulusBounds->rMax()) > m_tolerance)) {
+      ACTS_VERBOSE("Skip design '" << design << "' for element '"
+                                   << detElement->databaseEntryName());
+      continue;
+    }
+
+    radii.front() = annulusBounds->rMin();
+    radii.back() = annulusBounds->rMax();
+
+    ACTS_DEBUG("Accept design '" << design << "' for element '"
+                                 << detElement->databaseEntryName());
+
+    result.emplace();
+    const auto origValues = annulusBounds->values();
+
+    for (auto i = 0ul; i < radii.size() - 1; ++i) {
+      std::array<double, AnnulusBounds::eSize> values;
+      std::copy(origValues.begin(), origValues.end(), values.begin());
+      values[AnnulusBounds::eMinR] = radii[i];
+      values[AnnulusBounds::eMaxR] = radii[i + 1];
+
+      auto bounds = std::make_shared<AnnulusBounds>(values);
+
+      auto newDetElement =
+          GeoModelDetectorElement::createDetectorElement<DiscSurface>(
+              detElement->physicalVolume(), bounds, detElement->transform(gctx),
+              detElement->thickness());
+      auto newSurface = newDetElement->surface().getSharedPtr();
+
+      newDetElement->setDatabaseEntryName(detElement->databaseEntryName());
+
+      result->emplace_back(newSurface, newDetElement);
+    }
+
+    return result;
+  }
+
+  ACTS_WARNING("Could not split '" << detElement->databaseEntryName()
+                                   << "' (rmin: " << annulusBounds->rMin()
+                                   << ", rmax: " << annulusBounds->rMax()
+                                   << ")");
```

I guess this might be safer as an exception?

---

Hmm I thought in case there is no matching split pattern, we can fallback to use the unsplit surface (similar as if now `AnnulusBounds` were found.

---

I see. Generally, WARNINGs / silent errors are hard to figure out so that's why I'm wondering if this should be one.